lib/_repl.scm

```diff
@@ -4579,16 +4584,15 @@
 (define help-browser
   ##help-browser)
 
-(define-prim (##show-definition-of subject)
-  (let ((s
-         (cond ((##procedure? subject)
-                (##object->string (##procedure-name subject)))
-               (else
-                (##object->string subject)))))
-    (##show-help "Definition of " s)))
+(define-prim (##show-help-for-procedure proc)
```

Comment:
The concept index has many entries that do not correspond to Scheme objects. They can be general topics. What should we do about them?

Reply:
BTW `(help time)` should probably work since it's something newbies are likely to try. So `help` could be a macro, and it could treat `(quote foo)` and `foo` the same.

Reply:
If a help procedure is needed such that the argument comes from a run-time variable, could add a separate `help-proc` which is called by the macro

Reply:
Unfortunately, making `help` a macro would not work in these cases:

```
(define f car)
(help f)
(help (car (list bitwise-and bitwise-ior)))
```

This happens in higher-order code...  In other words, `(help foo)` does not mean "give me help on the thing called foo" but rather "give me help on the thing bound to foo" (which is normally a procedure, but could be extended at some point with other things that have documentation).  This is similar to `(trace E)` which means "add tracing to the  procedure that is the result of evaluating E".  For a user that doesn't like `help` being a procedure, they can get into the habit of always adding a quote in front of the parameter: `(help 'car)` and `(help 'time)`.

Reply:
The ideal situation would be to make `help` a macro if it could check if the parameter is the name of a macro. But currently this is not possible.

Reply:
If `help` is a procedure and `no-such` is not bound, does `(help no-such)` raise a predictable type of exception? If so, could `help` catch that exception and ask the user "did you mean `'foo`"?

Reply:
In evaluating `(help no-such)` the error is when evaluating `no-such`... the `help` procedure hasn't been called yet (and in fact it could be that the variable `help` isn't bound to the help procedure).

An alternative would be to define `help` as a macro such that when the parameter is a symbol it expands to an evaluation of the symbol (as a variable) guarded by an exception handler that detects when it is unbound.  If it is unbound then it could be treated as a quoted symbol (the name of a special form, a concept, etc), otherwise it is treated like currently.

Reply:
I'll prototype this idea...

Reply:
Yep, macro not procedure. Thanks for trying it out!

Reply:
I wrote the code below to demonstrate the idea.  Unfortunately it doesn't work when the parameter is the name of a macro because in that case the expansion of `(help time)` will contain a reference of the identifier `time` and the Scheme code parser (in `eval` and the compiler) will raise the compile-time exception `Macro name can't be used as a variable: time`. So there would need to be a macro-expansion-time mechanism to determine what an identifier is bound to, but that is too early in the evaluation pipeline. An alternative would be to add a special form to the interpreter and compiler: `(if-bound-to-macro? identifier then else)`, but this is a huge hack so I'll have to think about it...

```
(define-syntax help
  (lambda (src)
    (##deconstruct-call
     src
     2
     (lambda (arg-src)
       (let ((arg (##desourcify arg-src)))
         (if (symbol? arg)
             `(##help-on-name ',arg
                              (##with-exception-catcher
                               (lambda (exc)
                                 '#!unbound)
                               (lambda ()
                                 ,arg-src)))
             `(##help-on-value ,arg-src)))))))

(define (##help-on-name name #!optional (value '#!unbound))
  (##help-on-value
   (if (##unbound? value)
       name
       value)))
      
(define (##help-on-value value)
  ...) ;; same definition as current ##help procedure
```

Reply:
Makes sense. We can keep `help` as a procedure and revisit the topic if that macro system change is made in the future for other reasons.

Reply:
I think the right way would be to use `cond-expand` and to add a new feature tester for testing for the existence of a special form, e.g. `(special-form time)` in addition to the current `(library <lib>)` and `(compilation-target <target>)`.  So the test would look like this:

```
(##help-on-name 'time
                (##cond-expand ((special-form time) '#!unbound)
                               (else (##with-exception-catcher ...time...))))
```

Ideally this special form tester extension should eventually be available in other implementations of Scheme, so it would be good to poll implementors to see if this is an acceptable approach.  If not then Gambit could use the feature tester `(##special-form time)` to avoid clashing with a future extension of `cond-expand` where `(special-form <id>)` might mean something else.

Reply:
Another interesting poll would be to see what are all the current extensions of `cond-expand` that are implemented by other Scheme implementations.  It might lead to some standardization of feature tester names.  So for Gambit, `(compilation-target <target>)` is one such extension.

Reply:
Sounds good!

Reply:
I wonder if `cond-expand` should have a feature test to test the available feature tests :-)

Reply:
Ha! I hope `(cond-expand ((has-feature-tester has-feature-tester) #t) (else #f))` never returns `#f`!